coreneuron/io/setup_fornetcon.cpp

```diff
@@ -0,0 +1,220 @@
+  }
+
+  // How many NetCons (weight groups) are involved.
+  // Also count how many weight groups for each target instance.
+  // For the latter we can count in the dparam fornetcon slot.
+
+  // zero the dparam fornetcon slot for counting and count number of slots.
+  size_t n_perm_indices = 0;
+  for (auto it = type_to_slot.begin(); it != type_to_slot.end(); ++it) {
+    int mtype = it->first;
+    int fnslot = it->second;
+    int nodecount = nt._ml_list[mtype]->nodecount;
+    for (int i=0; i < nodecount; ++i) {
+      int* fn = fornetcon_slot(mtype, i, fnslot, nt);
+      *fn = 0;
+      n_perm_indices += 1;
+    }
+  }
+
+  // Count how many weight groups for each slot and total number of weight groups
+  size_t n_weight_perm = 0;
+  for (int i = 0; i < nt.n_netcon; ++i) {
+    NetCon& nc = nt.netcons[i];
+    int mtype = nc.target_->_type;
+    auto search = type_to_slot.find(mtype);
+    if (search != type_to_slot.end()) {
+      int i_instance = nc.target_->_i_instance;
+      int* fn = fornetcon_slot(mtype, i_instance, search->second, nt);
+      *fn += 1;
+      n_weight_perm += 1;
+    }
+  }
+
+  // Displacement vector has an extra element since the number for last item
+  // at n-1 is x[n] - x[n-1] and number for first is x[0] = 0.
+  nt._fornetcon_perm_indices.resize(n_perm_indices + 1);
+  nt._fornetcon_weight_perm.resize(n_weight_perm);
+
+  // From dparam fornetcon slots, compute displacement vector, and
+  // set the dparam fornetcon slot to the index of the displacement vector
+  // to allow later filling the _fornetcon_weight_perm.
+  size_t i_perm_indices = 0;
+  nt._fornetcon_perm_indices[0] = 0;
+  for (auto it = type_to_slot.begin(); it != type_to_slot.end(); ++it) {
+    int mtype = it->first;
+    int fnslot = it->second;
+    int nodecount = nt._ml_list[mtype]->nodecount;
+    for (int i=0; i < nodecount; ++i) {
+      int* fn = fornetcon_slot(mtype, i, fnslot, nt);
+      nt._fornetcon_perm_indices[i_perm_indices + 1] =
+        nt._fornetcon_perm_indices[i_perm_indices] + size_t(*fn);
+      *fn = int(nt._fornetcon_perm_indices[i_perm_indices]);
+      i_perm_indices += 1;
+    }
+  }
+
+  // One more iteration over NetCon to fill in weight index for
+  // nt._fornetcon_weight_perm. To help with this we increment the
+  // dparam fornetcon slot on each use.
+  for (int i = 0; i < nt.n_netcon; ++i) {
+    NetCon& nc = nt.netcons[i];
+    int mtype = nc.target_->_type;
+    auto search = type_to_slot.find(mtype);
+    if (search != type_to_slot.end()) {
+      int i_instance = nc.target_->_i_instance;
+      int* fn = fornetcon_slot(mtype, i_instance, search->second, nt);
+      size_t nc_w_index = size_t(nc.u.weight_index_);
+      nt._fornetcon_weight_perm[size_t(*fn)] = nc_w_index;
+      *fn += 1; // next item conceptually adjacent
+    }
+  }
+
+  // Put back the proper values into the dparam fornetcon slot
+  i_perm_indices = 0;
+  for (auto it = type_to_slot.begin(); it != type_to_slot.end(); ++it) {
```

Comment:
nitpick: 
  for (const auto& kv : type_to_slot) {
  }

then use : `kv.first` and `kv.second`